### DISCUSSION TITLE
Validating mutually exclusive fields with Yup

### DISCUSSION THREAD

**Discussion by alessandrojcm** [react-hook-form/react-hook-form#4687] [Q&A] [answered]
Hi there, I've been struggling to implement this form at work for quite a few days now; but I can't seem to have it working. We have a form kinda like this (simplified a bit for the sake of clarity):

* There are 4 number inputs
* Each input is mutually exclusive
* If the user inputs a number in one of the fields, that field should be marked as required and the validation should be turned on only for that field (validation criteria being only numbers greater than 0)
* The other fields should be then marked as optional and their values (if any) transformed to 0

This is the yup schema I have come with; mi idea is to have an input of type `hidden` on my form to control the active rate, so each time the user types at an input `activeRate` gets changed to that input name and the schema validation gets triggered:

```javascript
yup.object().shape({
  activeRate: yup
    .string()
    .oneOf(["cost1", "cost2", "cost3", "cost4"])
    .default("cost1")
    // This is only for controlling the form, so we strip it from the final value
    .strip(true),
  cost1: yup.number().when("activeRate", {
    is: "cost1",
    then: yup.number().required().default(0).positive().typeError(),
    otherwise: yup
      .number()
      .optional()
      .transform(() => 0)
  }),
  cost2: yup.number().when("activeRate", {
    is: "cost2",
    then: yup.number().required().default(0).positive().typeError(),
    otherwise: yup
      .number()
      .notRequired()
      .transform(() => 0)
  }),
  cost3: yup.number().when("activeRate", {
    is: "cost3",
    then: yup.number().required().default(0).positive().typeError(),
    otherwise: yup
      .number()
      .notRequired()
      .transform(() => 0)
  }),
  cost4: yup.number().when("activeRate", {
    is: "cost4",
    then: yup.number().required().default(0).positive().typeError(),
    otherwise: yup
      .number()
      .notRequired()
      .transform(() => 0)
  })
});
```

But, as you can see in [this codesandbox](https://codesandbox.io/s/cool-lake-741m7?fontsize=14&hidenavigation=1&theme=dark), `activeRate` doesn't change and, more importantly, the number values of the inputs don't change either; what could I be missing here?

**Comment by jorisre** [comment, marked as answer]
Hi @alessandrojcm 

I think I got it working: https://codesandbox.io/s/pensive-dust-qfpzd
I've added error's message under fields for more clarity.

It looks like you can't watch a value with `when` and `strip` at the time with Yup. I removed `strip` to make it works.
As an alternative, you can exclude the `activeRate` value on submit:
```tsx
const submit = (data) => {
    const { activeRate, ...values } = data;
    doSomething(values)
  };
```

**Reply by alessandrojcm** [reply]
Brilliant @jorisre, thanks a million! Just another question: from what I see, `transform` only gets applied after submission, is that right?

**Reply by jorisre** [reply]
According to the yup doc: https://github.com/jquense/yup#mixedtransformcurrentvalue-any-originalvalue-any--any-schema

So, each time the resolver is called, it depends on the `mode` you are using 'all | onChange | onBlur | ...'

If you're curious, here is the yup resolver: https://github.com/react-hook-form/resolvers/blob/master/yup/src/yup.ts#L48

**Reply by steveoh** [reply]
any chance that code sandbox can be restored?

**Comment by FooBarRaz** [comment]
Here's how I accomplished this: 

```
const schema = yup.object().shape({
    someField: yup.string()
        .when('someOtherField', (someOtherFieldValue, thisField) =>
            someOtherFieldValue ?
                thisField.oneOf([null, undefined, ''], 'someField  and someOtherField cannot be specified together')
                : thisField.required()),
    someOtherField: yup.string().notRequired()
})
